maplibre/src/coords.rs

```diff
@@ -22,54 +22,50 @@ pub const EXTENT_UINT: u32 = 4096;
 pub const EXTENT_SINT: i32 = EXTENT_UINT as i32;
 pub const EXTENT: f64 = EXTENT_UINT as f64;
 pub const TILE_SIZE: f64 = 512.0;
-pub const MAX_ZOOM: usize = 32;
 
-// FIXME: MAX_ZOOM is 32, which means max bound is 2^32, which wouldn't fit in u32 or i32
-// Bounds are generated 0..=31
-pub const ZOOM_BOUNDS: [u32; MAX_ZOOM] = create_zoom_bounds::<MAX_ZOOM>();
+/// Maximal zoom level. Typically, the maximal zoom is 18 as described
+/// [here](https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Zoom_levels).
+/// This implementation allows 30 because this covers all tile sizes that can be represented
+/// with signed 32-bit integers.
+pub(crate) const MAX_ZOOM: u8 = 30;
 
-const fn create_zoom_bounds<const DIM: usize>() -> [u32; DIM] {
-    let mut result: [u32; DIM] = [0; DIM];
-    let mut i = 0;
-    while i < DIM {
-        result[i] = 2u32.pow(i as u32);
-        i += 1;
-    }
-    result
-}
-
-/// Represents the position of a node within a quad tree. The first u8 defines the `ZoomLevel` of the node.
-/// The remaining bytes define which part (north west, south west, south east, north east) of each
-/// subdivision of the quadtree is concerned.
-///
-/// TODO: We can optimize the quadkey and store the keys on 2 bits instead of 8
+/// Represents the position of a node within a quad tree using the zoom level and morton encoding.
 #[derive(Ord, PartialOrd, Eq, PartialEq, Clone, Copy)]
-pub struct Quadkey([ZoomLevel; MAX_ZOOM]);
+pub struct Quadkey {
+    zoom_level: ZoomLevel,
+    encoded_coords: u64
+}
 
 impl Quadkey {
-    pub fn new(quad_encoded: &[ZoomLevel]) -> Self {
-        let mut key = [ZoomLevel::default(); MAX_ZOOM];
-        key[0] = (quad_encoded.len() as u8).into();
-        for (i, part) in quad_encoded.iter().enumerate() {
-            key[i + 1] = *part;
+    pub fn new(zoom_level: ZoomLevel, x: u32, y: u32) -> Self {
+        // use morton enconding / Z-ordering to build linear quadtree index
+        let encoded_coords = morton_encoding::morton_encode([x, y]);
```

Comment:
Interesting approach. As this is used a lot, can we benchmark the encoding/decoding OR do you have a guess on how fast this is?

Reply:
It compiles to ~20 branchless instructions of x86 and should be really fast: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=6693a9f5ac38027b1fefe470f4208428 (select ASM in the ... menu on the top left)